### PR TITLE
[SPARK-32456][SS][FOLLOWUP] Update doc to note about using SQL statement with streaming Dataset

### PR DESCRIPTION
**Pull request by HeartSaVioR**
### What changes were proposed in this pull request?

This patch proposes to update the doc (both SS guide doc and Dataset dropDuplicates method doc) to leave a note to check on using SQL statements with streaming Dataset.

Once end users create a temp view based on streaming Dataset, they won't bother with thinking about "streaming" and do whatever they do with batch query. In many cases it works, but not just smoothly for the case when streaming aggregation is involved. They still need to concern about maintaining state store.

### Why are the changes needed?

Although SPARK-32456 fixed the weird error message, as a side effect some operations are enabled on streaming workload via SQL statement, which is error-prone if end users don't indicate what they're doing.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Only doc change.

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
I'm marking this as draft as I'd like to see which is preferred - just document to warn about end users (this PR) vs collect and prevent some error-prone operations for streaming workload (proposed https://github.com/apache/spark/pull/29256#discussion_r471945148). If we don't mind covering this with guide doc, this PR can be converted to "ready-to-review".

**Comment by HeartSaVioR** [comment]
cc. @cloud-fan @xuanyuanking @tdas @zsxwing @gaborgsomogyi

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 1]
```diff
@@ -2525,14 +2525,19 @@ class Dataset[T] private[sql](
...
+   * slightly different behavior. (see below)
    *
    * For a static batch [[Dataset]], it just drops duplicate rows. For a streaming [[Dataset]], it
    * will keep all data across triggers as intermediate state to drop duplicates rows. You can use
    * [[withWatermark]] to limit how late the duplicate data can be and system will accordingly limit
    * the state. In addition, too late data older than watermark will be dropped to avoid any
    * possibility of duplicates.
    *
+   * Note that for a streaming [[Dataset]], this method only returns distinct rows only once,
+   * regardless of the output mode. Spark may convert the `distinct` operation to aggregation`,
```
Comment (gaborgsomogyi) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2538]: > Spark may convert the `distinct` operation to aggregation`

Can we add an example to the user? This question applies to other places.

**Review by gaborgsomogyi — COMMENTED** [review]
Only a minor question.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 1 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2538]: Probably I need to reword, as the behavior is tied to the catalyst rules, which can be changed in anytime. It would be just enough that `, which the behavior may not be same with using `distinct` in SQL statement`. I'll fix it.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 2]
```diff
@@ -2525,14 +2525,19 @@ class Dataset[T] private[sql](
...
+   * slightly different behavior. (see below)
    *
    * For a static batch [[Dataset]], it just drops duplicate rows. For a streaming [[Dataset]], it
    * will keep all data across triggers as intermediate state to drop duplicates rows. You can use
    * [[withWatermark]] to limit how late the duplicate data can be and system will accordingly limit
    * the state. In addition, too late data older than watermark will be dropped to avoid any
    * possibility of duplicates.
    *
+   * Note that for a streaming [[Dataset]], this method only returns distinct rows only once
+   * regardless of the output mode, which the behavior may not be same with using distinct in
```
Comment (xuanyuanking) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2538]: +1 for the second version, actually the original comments also confusing, we need to emphasize the distinct clause in SQL.

File: docs/structured-streaming-programming-guide.md [hunk 3]
```diff
@@ -861,6 +861,10 @@ isStreaming(df)
 </div>
 </div>
 
+You may want to check the logical plan of the query, as Spark converts the operation into another operation, which includes adding streaming aggregation. (e.g. count, distinct, union, etc.)
```
Comment (xuanyuanking) [inline docs/structured-streaming-programming-guide.md:864]: I think the operation converting is internal behavior, maybe it's not clear enough for asking the end-user to check it. How about we just comment on the behavior difference between SQL distinct and dataset dropDuplicate, WDYT?

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 1 above)
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: +1 for the reword version, actually the original comments also confusing, we need to emphasize the distinct clause in SQL.

File: docs/structured-streaming-programming-guide.md (hunk 3 above)
Comment (HeartSaVioR) [reply docs/structured-streaming-programming-guide.md:864]: The thing is whether Spark injects streaming aggregation which end users have to maintain or not, and that can be checked by looking into logical plan, right? I didn't mean they need to find the distinct in logical plan and how Spark changes the operation. They just need to check for stateful operations.

SQL distinct and Dataset dropDuplicate aren't the only difference. SQL union and Dataset union are also different. The cases can increase and decrease according to the Spark catalyst rules, which is not we can ensure the doc be in sync.

Comment (HeartSaVioR) [reply docs/structured-streaming-programming-guide.md:864]: Probably we can reword here as well to simplify, like 

```
You may want to check the query plan of the query, as Spark could inject stateful operations during interpret of the query. Once stateful operations are injected in the query plan, you may need to check your query with considerations in stateful operations. (e.g. output mode, watermark, state store size maintenance, etc.)
```

If the reworded sentences sound better then I can update.

**Comment by HeartSaVioR** [comment]
retest this, please

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 1 above)
Comment (gaborgsomogyi) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: +1 to emphasize the distinct clause in SQL.

**Comment by gaborgsomogyi** [comment]
retest this please

**Comment by gaborgsomogyi** [comment]
The doc generation failed but seems unrelated:
```
java.lang.RuntimeException: Could not create directory /home/runner/work/spark/spark/external/avro/target/streams/$global/projectDescriptors/$global/streams
```

**Review by gaborgsomogyi — APPROVED** [review]
LGTM.

File: docs/structured-streaming-programming-guide.md (hunk 3 above)
Comment (cloud-fan) [reply docs/structured-streaming-programming-guide.md:864]: The reworded version looks better.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 4]
```diff
@@ -2525,14 +2525,19 @@ class Dataset[T] private[sql](
 
   /**
    * Returns a new Dataset that contains only the unique rows from this Dataset.
-   * This is an alias for `distinct`.
+   * This is an alias for `distinct` on batch [[Dataset]]. For streaming [[Dataset]], it would show
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2528]: It is an alias, `Dataset.distinct` calls `dropDuplicates`.

If we want to highlight the different between SQL distinct and `Dataset.distinct`, we should update document of `Dataset.distinct`.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 5]
```diff
@@ -2548,6 +2553,10 @@ class Dataset[T] private[sql](
    * the state. In addition, too late data older than watermark will be dropped to avoid any
    * possibility of duplicates.
    *
+   * Note that for a streaming [[Dataset]], this method only returns distinct rows only once
+   * regardless of the output mode, which the behavior may not be same with using distinct in
+   * SQL statement against streaming [[Dataset]].
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2558]: ditto, there is no `dropDuplicates` in SQL. If we want to highlight differences between SQL API and Dataset API, let's update the doc of `Dataset.distinct`

File: docs/structured-streaming-programming-guide.md (hunk 3 above)
Comment (HeartSaVioR) [reply docs/structured-streaming-programming-guide.md:864]: Thanks for the input. Will update.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 6]
```diff
@@ -3131,8 +3131,12 @@ class Dataset[T] private[sql](
    * Returns a new Dataset that contains only the unique rows from this Dataset.
    * This is an alias for `dropDuplicates`.
    *
+   * Note that for a streaming [[Dataset]], this method only returns distinct rows only once
+   * regardless of the output mode, which the behavior may not be same with `DISTINCT` in SQL
+   * against streaming [[Dataset]].
+   *
    * @note Equality checking is performed directly on the encoded representation of the data
-   * and thus is not affected by a custom `equals` function defined on `T`.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:3135]: unnecessary change?

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 7]
```diff
@@ -3131,8 +3131,12 @@ class Dataset[T] private[sql](
    * Returns a new Dataset that contains only the unique rows from this Dataset.
    * This is an alias for `dropDuplicates`.
    *
+   * Note that for a streaming [[Dataset]], this method only returns distinct rows only once
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:3134]: `only returns distinct rows only once` -> `returns distinct rows only once`

**Comment by cloud-fan** [comment]
github action passed, merging to master, thanks!

**Comment by HeartSaVioR** [comment]
Thanks all for reviewing and merging!
